File: samples/python2/lk_track.py
import numpy as np
import cv2, cv
import video
from common import anorm2, draw_str
from time import clock

help_message = '''
USAGE: lk_track.py [<video_source>]

Keys:
  1 - toggle old/new CalcOpticalFlowPyrLK implementation
  SPACE - reset features
'''




lk_params = dict( winSize  = (3, 3), 
                  maxLevel = 2, 
                  criteria = (cv2.TERM_CRITERIA_EPS | cv2.TERM_CRITERIA_COUNT, 10, 0.03),
                  derivLambda = 0.0 )    

feature_params = dict( maxCorners = 1000, 
                       qualityLevel = 0.1,
                       minDistance = 5,
                       blockSize = 5 )

def calc_flow_old(img0, img1, p0):
    p0 = [(x, y) for x, y in p0.reshape(-1, 2)]
    h, w = img0.shape[:2]
    img0_cv = cv.CreateMat(h, w, cv.CV_8U)
    img1_cv = cv.CreateMat(h, w, cv.CV_8U)
    np.asarray(img0_cv)[:] = img0
    np.asarray(img1_cv)[:] = img1
    t = clock()
    features, status, error  = cv.CalcOpticalFlowPyrLK(img0_cv, img1_cv, None, None, p0, 
        lk_params['winSize'], lk_params['maxLevel'], (cv.CV_TERMCRIT_EPS | cv.CV_TERMCRIT_ITER, 10, 0.03), 0, p0)
    return np.float32(features), status, error, clock()-t

def main():
    import sys
    try: video_src = sys.argv[1]
    except: video_src = video.presets['chess']

    cv2.namedWindow('img', 0)

    track_len = 4
    tracks = []
    cam = video.create_capture(video_src)
    old_mode = True
    while True:
        ret, frame = cam.read()
        vis = frame.copy()
        if len(tracks) > 0:
            p0 = np.float32([tr[-1] for tr in tracks]).reshape(-1, 1, 2)
            img0 = cv2.cvtColor(prev_frame, cv.CV_BGR2GRAY)
            img1 = cv2.cvtColor(frame, cv.CV_BGR2GRAY)
            if old_mode:
                p1,  st, err, dt = calc_flow_old(img0, img1, p0)
            else:
                t = clock()
                p1,  st, err = cv2.calcOpticalFlowPyrLK(img0, img1, p0, **lk_params)
                dt = clock()-t
            for tr, (x, y) in zip(tracks, p1.reshape(-1, 2)):
                tr.append((x, y))
                if len(tr) > 10:
                    del tr[0]
                cv2.circle(vis, (x, y), 2, (0, 255, 0), -1)
            cv2.polylines(vis, [np.int32(tr) for tr in tracks], False, (0, 255, 0))
            draw_str(vis, (20, 20), ['new', 'old'][old_mode]+' mode')
            draw_str(vis, (20, 40), 'time: %.02f ms' % (dt*1000))
        prev_frame = frame.copy()

        cv2.imshow('img', vis)
        ch = cv2.waitKey(5)
        if ch == 27:
            break
        if ch == ord(' ') or len(tracks) == 0:
            gray = cv2.cvtColor(frame, cv.CV_BGR2GRAY)
            p = cv2.goodFeaturesToTrack(gray, **feature_params)
            p = [] if p is None else p.reshape(-1, 2)
            tracks = []
            for x, y in np.float32(p):
                tracks.append([(x, y)])
        if ch == ord('1'):
            old_mode = not old_mode

if __name__ == '__main__':
    main()

File: samples/cpp/lk_track.cpp
#include <opencv2/highgui/highgui.hpp>
#include <opencv2/core/core.hpp>
#include <opencv2/imgproc/imgproc.hpp>
#include <opencv2/video/video.hpp>
#include <iostream>

using namespace std;
using namespace cv;


int main(int argc, char * argv[])
{
  VideoCapture cap;
  if(argc>1)
  {
    string filename = argv[1];
    cap.open(filename);
  }else
  {
    cap.open(0);
  }


   Mat img, mask, vis;
   Mat gray, prevGray;
   vector <Point2f> p, p0, p0r, p1;
   vector < vector<Point2f> > tracks, new_tracks;
   vector <float> d;
   vector <Point2f> tmp;
   vector <uchar> status;
   vector <float> err;
   Size winSize(15, 15);
   int maxLevel = 3;
   TermCriteria criteria(CV_TERMCRIT_EPS | CV_TERMCRIT_ITER, 20, 0.03);
   vector <bool> good;
   int track_len = 10;
   int maxCorners = 500;
   double qualityLevel = 0.3;
   double minDistance = 7;
   int blockSize = 7;
   for(;;)
   {
     cap>>img;
     resize(img,img, Size(720,568));
     if(img.empty()) continue;
     cvtColor(img,gray, COLOR_BGR2GRAY);
     break;
   }
   gray.copyTo(prevGray);
   tracks.clear();
   p1.clear();
   p0r.clear();

   for(;;)
   {
     cap>>img;
     resize(img,img, Size(720,568));
     img.copyTo(vis);
     cvtColor(img,gray, COLOR_BGR2GRAY);
     gray.copyTo(mask);
     mask = Scalar(255);
     if(!tracks.empty())
     {
       for(int i = 0; i < (int)tracks.size(); i++)
       {
         circle(mask, tracks[i].back(), 5, Scalar(0), -1);
       }
     }
     p.clear();
     goodFeaturesToTrack(gray, p, maxCorners, qualityLevel, minDistance, mask, blockSize);

     for(int i = 0; i < (int)p.size(); i++)
     {
       tmp.clear();
       tmp.push_back(p[i]);
       tracks.push_back(tmp);
     }
     if(tracks.size() > 0)
     {
       p0.clear();
       for(int i = 0; i<(int)tracks.size(); i++)
       {
         p0.push_back(tracks[i].back());
       }

       calcOpticalFlowPyrLK(prevGray, gray, p0, p1, status, err, winSize, maxLevel, criteria, 0, 0.001);
       calcOpticalFlowPyrLK(gray, prevGray, p1, p0r, status, err, winSize, maxLevel, criteria, 0, 0.001);
       good.clear();
       for(int i = 0; i <(int) p1.size(); i++)
       {
          float dx = abs(p0[i].x - p0r[i].x);
          float dy = abs(p0[i].y - p0r[i].y);
          good.push_back(std::max(dx,dy) < 1);
       }
       new_tracks.clear();
       for( int i = 0; i <(int) tracks.size(); i++)
       {
          if(!good[i])
          {
             continue;
          }
          tracks[i].push_back(p1[i]);

          if((int)tracks[i].size() > track_len)
          {
             // remove first element
             tracks[i].erase(tracks[i].begin());
          }
          new_tracks.push_back(tracks[i]);
          circle(vis, tracks[i].back(), 2, Scalar(0,255,0), -1);
       }

       std::swap(tracks, new_tracks);
       new_tracks.clear();
       for (int i = 0; i < (int)tracks.size(); i ++)
       {
          vector <Point2i> dst;
          dst.clear();
          std::copy(tracks[i].begin(), tracks[i].end(), std::back_inserter(dst));
          polylines(vis,dst,false,Scalar(0,255,0));
       }
     }
     gray.copyTo(prevGray);

     imshow("IMG",vis);
     int key = waitKey(1);
	 if(key == 27)
		 break;

}
  return 0;
}
